runtime/src/iree-amd-aie/driver/xrt/direct_command_buffer.cc

```diff
@@ -0,0 +1,446 @@
+    } catch (...) {
+      IREE_TRACE_ZONE_END(z0);
+      return iree_make_status(
+          IREE_STATUS_RESOURCE_EXHAUSTED,
+          "could not allocate memory for sub buffers being binded");
+    }
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_xrt_direct_command_buffer_dispatch(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_executable_t* executable, int32_t entry_point,
+    uint32_t workgroup_x, uint32_t workgroup_y, uint32_t workgroup_z) {
+  iree_hal_xrt_direct_command_buffer_t* command_buffer =
+      iree_hal_xrt_direct_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+  // Lookup kernel parameters used for side-channeling additional launch
+  // information from the compiler.
+  iree_hal_xrt_kernel_params_t kernel_params;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_xrt_native_executable_entry_point_kernel_params(
+              executable, entry_point, &kernel_params));
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_resource_set_insert(command_buffer->resource_set, 1,
+                                       &executable));
+  xrt::kernel kernel = *kernel_params.kernel;
+  xrt::bo instr = *kernel_params.instr;
+  uint32_t num_instr = kernel_params.num_instr;
+
+  xrt::run run = xrt::run(kernel);
+
+  // index to push arguments on the kernel;
+  iree_host_size_t arg_index = 0;
+
+  // first argument is the LX6 instructions
+  instr.sync(XCL_BO_SYNC_BO_TO_DEVICE);
+  run.set_arg(arg_index++, instr);
+
+  // second argument is the number of LX6 instructions
+  run.set_arg(arg_index++, num_instr);
+
+
+  // collect all the other bindings in a list
+  std::vector<xrt::bo*> args_buffer;
+  // Copy descriptors from all sets to the end of the current segment for later
+  // access.
+  iree_host_size_t set_count =
+      iree_hal_xrt_pipeline_layout_descriptor_set_count(kernel_params.layout);
+  command_buffer->set_count = set_count;
+  for (iree_host_size_t i = 0; i < set_count; ++i) {
+    // TODO: cache this information in the kernel info to avoid recomputation.
+    iree_host_size_t binding_count =
+        iree_hal_xrt_descriptor_set_layout_binding_count(
+            iree_hal_xrt_pipeline_layout_descriptor_set_layout(
+                kernel_params.layout, i));
+    iree_host_size_t base_index =
+        iree_hal_xrt_pipeline_layout_base_binding_index(kernel_params.layout,
+                                                        i);
+    command_buffer->descriptor_sets[i].binding_count = binding_count;
+    for (iree_host_size_t j = 0; j < binding_count; ++j) {
+      xrt::bo* arg_buffer = command_buffer->descriptor_sets[i].bindings[j];
+      args_buffer.push_back(arg_buffer);
+      arg_buffer->sync(XCL_BO_SYNC_BO_TO_DEVICE);
+      run.set_arg(arg_index + base_index + j,
+                  *arg_buffer);
+    }
+  }
+  run.start();
+  run.wait();
+
+  instr.sync(XCL_BO_SYNC_BO_FROM_DEVICE);
```

Comment:
Hmm, I'm not sure we need all of these sync calls.. Let's chat later about this.